File: backend/app/schemas/stats.py
from pydantic import BaseModel


class EmployeeStats(BaseModel):
    employee_id: int
    employee_name: str
    role: str
    shift_counts: dict[str, int]
    total_work_days: int
    total_off_days: int
    total_special_days: int
    weekend_work_count: int
    max_consecutive_work: int
    consecutive_off_count: int
    preferred_satisfied: int


class ShiftStats(BaseModel):
    total: int
    per_day_avg: float
    min: int
    max: int


class DailyCoverageStats(BaseModel):
    day: int
    weekday: int
    A: int
    B: int
    C: int
    D: int
    M: int
    total: int


class MonthSummary(BaseModel):
    year: int
    month: int
    num_days: int
    total_shifts: int
    solve_status: str
    objective_value: float | None
    solve_time: float | None


class MonthStatsResponse(BaseModel):
    month_summary: MonthSummary
    per_employee: list[EmployeeStats]
    per_shift: dict[str, ShiftStats]
    daily_coverage: list[DailyCoverageStats]
    weekend_summary: dict
    violations_summary: dict
    soft_constraint_stats: dict | None

File: frontend/src/api/stats.ts
import { apiClient } from './client'

export interface EmployeeStats {
  employee_id: number
  employee_name: string
  role: string
  shift_counts: Record<string, number>
  total_work_days: number
  total_off_days: number
  total_special_days: number
  leave_type_counts: Record<string, number>
  total_leave_days: number
  weekend_work_count: number
  max_consecutive_work: number
  consecutive_off_count: number
  preferred_satisfied: number
}

export interface ShiftStats {
  total: number
  per_day_avg: number
  min: number
  max: number
}

export interface DailyCoverageStats {
  day: number
  weekday: number
  A: number
  B: number
  C: number
  D: number
  M: number
  total: number
}

export interface MonthStats {
  month_summary: {
    year: number
    month: number
    num_days: number
    total_shifts: number
    solve_status: string
    objective_value: number | null
    solve_time: number | null
  }
  per_employee: EmployeeStats[]
  per_shift: Record<string, ShiftStats>
  daily_coverage: DailyCoverageStats[]
  weekend_summary: {
    saturdays: { day: number; workers: string[]; shifts: string[] }[]
    sundays: { day: number; workers: string[]; shifts: string[] }[]
  }
  violations_summary: {
    total_violations: number
    hard_violations: number
    soft_warnings: number
    is_valid: boolean
  }
  soft_constraint_stats: Record<string, number> | null
  leave_types: { code: string; name: string; color_bg: string; color_text: string }[]
}

export async function getMonthStats(year: number, month: number): Promise<MonthStats> {
  const { data } = await apiClient.get<MonthStats>(`/stats/${year}/${month}`)
  return data
}
